_static/js/citesupport-es6.js

```diff
@@ -0,0 +1,178 @@
+
+class CiteSupport {
+
+    constructor() {
+        this.config = {
+            debug: false,
+            mode: 'note',
+            defaultLocale: 'en-US',
+            defaultStyle: 'jm-indigobook-law-review',
+            posToCitationId: {},
+            citationIdToPos: {},
+            processorReady: false,
+            citationByIndex: [],
+            current: {},
+        };
+    }
+
+    /**
+     * Logs messages to the console if `config.debug` is true
+     * @param  {string} txt The message to log
+     * @return {void}
+     */
+    debug(txt) {
+        if (this.config.debug) {
+            console.log(`*** ${txt}`);
+        }
+    }
+
+    citationAddOrEditHandler(e) {
+        this.debug('citationAddOrEditHandler()');
+
+        const menu = document.getElementById('cite-menu');
+        const citationItems = this.getCitationIdsFrom(menu);
+        const hasCitation = this.menuHasCitation(menu);
+        this.config.current = this.getCurrentCitationInfo();
+
+        if (citationItems.length === 0) {
+            if (hasCitation) {
+
+            }
+        }
+
+    }
+
+
+    /**
+     * Receives an HTML element as input. An HTML Collection is extracted from
+     *   the input element containing all child input[type="checkbox"] elements. The
+     *   HTML Collection is then iterated and the values of the checked boxes are
+     *   pushed to `citationItems` in the form of { id: `value`}. The "value" of
+     *   the checkbox element in this case is the citation ID.
+     *
+     * NOTE: This is a helper method used for this example and is NOT REQUIRED.
+     *
+     * @param  {HTMLElement} element An HTML element containing HTML input elements as children.
+     * @return {Object[]}      An array of objects containing citation IDs.
+     */
+    getCitationIdsFrom(element) {
+        this.debug('getCitationItemIdsFrom()');
+        const citationItems = [];
+        const checkboxes = element.getElementsByTagName('input');
+
+        for (const checkbox of checkboxes) {
+            if (checkbox.checked) {
+                citationItems.push({
+                    id: checkbox.value,
+                });
+            }
+        }
+
+        return citationItems;
+    }
+
+    /**
+     * Helper function used for getting information on the currently selected citation
+     *   menu object.
+     *
+     * NOTE: This is a helper method used for this example and is NOT REQUIRED.
+     *
+     * @return {Object} `info` object (defined below).
+     */
+    getCurrentCitationInfo() {
+        this.debug('getCurrentCitationInfo()');
+
+        /**
+         * Citation info object
+         * @type {Object}
+         * @prop {number} pos  - The position of the currently selected citation menu.
+         * @prop {string} citationID  - The `id` of the span element which holds the
+         *   inline citations.
+         * @prop {number} citationIndex  - I have no idea what this is for. FIXME
```

Comment:
When a citation is inserted, edited or removed, we will send a call to the processor with some data reflecting the current document state:
```
  @citation {Object} - data for the current citation
  @citationsPre {Array} - an array of previous citationID/noteIndex pairs
  @citationsPost {Array} - an array of following citationID/noteIndex pairs
```
* To collect the data from the document, we need to identify the current position in the list of citations. The `citationIndex` value here does that for us.
* The `citationID` is assigned by the processor. If we are editing an existing citation, it will be stamped with and ID, and we pick it up here for inclusion on `@citation`. If we are inserting a new citation, this will be a nil value.
* The `pos` value is a hack. It's the "chevron" position, which I've used as the hanger for the insert menu. That was a bad decision, and a source of confusion. If the insert menu were hung off of a (possibly blank, freshly inserted) citation object, this value would be unnecessary, and `citationIdToPos/posToCitationId` could be replaced with a simple state object that tells whether a given citationID exists or not (that becomes useful later, when the list of updates with a full set of citationIDs comes back from the processor).

Hope that makes sense ...